src/core/drive/page_renderer.ts

```diff
@@ -1,12 +1,10 @@
 import { Renderer } from "../renderer"
 import { PageSnapshot } from "./page_snapshot"
 import { ReloadReason } from "../native/browser_adapter"
-import { activateScriptElement, waitForLoad, nextEventLoopTick, getBodyElementId } from "../../util"
+import { activateScriptElement, waitForLoad, getBodyElementId } from "../../util"
 
 export class PageRenderer extends Renderer<HTMLBodyElement, PageSnapshot> {
-  static async renderElement(currentElement: HTMLBodyElement, newElement: HTMLBodyElement) {
-    await nextEventLoopTick()
```

Comment:
@agrobbin reverting these lines was necessary to pass CI. If they were crucial parts of the feature, could you try re-introducing them with test coverage that passes in their presence but fails in their absence?

Reply:
@seanpdoyle this was where I was struggling in #627. The `await nextEventLoopTick()` was necessary for reasons I explained in https://github.com/hotwired/turbo/pull/627#discussion_r921455484 (and are further described in https://github.com/hotwired/turbo/pull/305#issuecomment-1183505367), but for reasons unknown to me, it caused a bunch of other test failures that I could not understand.

Reply:
Thank you for sharing that context.

This might leave the release candidate in an awkward place: the CI failures introduced by the original changeset appear to be genuine, so the fact that they're failing signals that Real rendering regressions might have been released.

On the flip side, making CI pass (including the failing rendering tests and the test coverage introduced by #627) _without_ the `nextEventLoopTick()` signals that passing CI has the potential to break the feature introduced by #627 in a way that isn't covered by the test suite.

@agrobbin are you able to investigate an alternative pull request that maintains the changes you've introduced, exercises them with text coverage, and passes the rest of the suite?

Reply:
@agrobbin alternatively, if you can introduce a failing test that exhibits the behavior you've outlined in https://github.com/hotwired/turbo/pull/305#issuecomment-1183505367 and https://github.com/hotwired/turbo/pull/627#discussion_r921455484, I could troubleshoot on my own.

Reply:
I was looking for some help with that exact thing in https://github.com/hotwired/turbo/pull/627#issuecomment-1212647667, as I was not able to figure out why that particular addition failed tests, when things _did_ seem to work as expected when testing it locally.

Reply:
@agrobbin I've pushed up a second commit to try and restore the asynchronous rendering by moving the call to the general purpose renderer. Since this isn't covered by tests, it's difficult to verify on my own that it works as expected. Could you pull these changes down locally and attempt to reproduce the buggy behavior you've described in https://github.com/hotwired/turbo/pull/627#discussion_r921455484?

Reply:
@seanpdoyle, on it, give me a few minutes and I'll confirm one way or the other!

Reply:
It seems to work @seanpdoyle!

Reply:
Your change also seems to be way more obvious and less magical than mine, so thank you for that.

Reply:
awesome! how did you verify that it worked? Could you share an outline of the steps you took, or share some pseudocode to exercise it?

Reply:
It was a bit cumbersome, since there is no single compiled distrubtion of your changes, so what I did was add this resolution to my `package.json`:

```json
"@hotwired/turbo": "7.2.0-rc.1"
```

And then I explicitly edited the `dist/turbo.es2017-esm.js` file to match these changes. That allowed me to run this exact code on my application that was previously leveraging a patched version of the changes in #627. From there I could try it with `await nextAnimationFrame()` and without, and confirm that it does not work when it's missing, and that it does work when it's present. It *not* working manifests as the page caches being out of sync w/ the URL that is shown in the browser address bar when navigating forward and backward in browser history (after clicking on a few different links to navigate through several pages).

Reply:
Could you share more details specific about your application code, or could you generalize it to pseudocode or elements that could be added directly to the codebase's test suite?

Reply:
Let me try and boil it down to the basics that I can share. It might take me a little bit while I juggle some family stuff this afternoon.

Reply:
OK, here goes...

My application has a layout that looks like what I shared below:

```html
<html>
  <head>
    <meta name="turbo-body" content="app">
  </head>
  <body>
    <div id="app"></div>
  </body>
</html>
```

Within the `#app` element, the vast majority of links run through Turbo Drive. Imagine a few navigational links like this:

```html
<a href="/users">Users</a>
<a href="/projects">Projects</a>
<a href="/billing">Billing</a>
```

Here is some sample content for each of those pages, wrapped inside of the `#app` element:

```html
<!-- /users -->
<h1>Users</h1>
<p>You are the only user.</p>

<!-- /projects -->
<h1>Projects</h1>
<p>There are currently no projects.</p>

<!-- /billing -->
<h1>Billing</h1>
<p>You owe us $1.</p>
```

Clicking on each navigational link works as you'd expect through Turbo Drive, replacing the content within `#app`, updating `<head>` appropriately, etc.

If you start at Users, go to Projects, and then to Billing, everything is A-OK. When you start clicking the browser back/forward buttons however, the content within `#app` that is presented to you does not match the URL you see. Clicking the back button once will update the URL in the address bar from `/billing` to `/projects`, but continue to show the content from `/billing`. Clicking back again updates the URL to `/users`, but shows the content from `/projects`. Essentially, everything is off by 1.

That's the minimal set of changes that I think represent the situation that I experience when the `await nextAnimationFrame()` is not included.

I hope that's helpful!

Reply:
I tried to reproduce that locally with this diff:

```diff
diff --git a/src/core/view.ts b/src/core/view.ts
index e02835b..0c34160 100644
--- a/src/core/view.ts
+++ b/src/core/view.ts
@@ -3,7 +3,6 @@ import { Renderer, Render } from "./renderer"
 import { Snapshot } from "./snapshot"
 import { Position } from "./types"
 import { getAnchor } from "./url"
-import { nextAnimationFrame } from "../util"
 
 export interface ViewRenderOptions<E> {
   resume: (value: any) => void
@@ -127,7 +126,6 @@ export abstract class View<
   }
 
   async renderSnapshot(renderer: R) {
-    await nextAnimationFrame()
     await renderer.render()
   }
 
diff --git a/src/tests/functional/drive_custom_body_tests.ts b/src/tests/functional/drive_custom_body_tests.ts
index 54fe5a5..c418a7e 100644
--- a/src/tests/functional/drive_custom_body_tests.ts
+++ b/src/tests/functional/drive_custom_body_tests.ts
@@ -1,6 +1,6 @@
 import { test } from "@playwright/test"
 import { assert } from "chai"
-import { nextBody, pathname, getFromLocalStorage, setLocalStorageFromEvent } from "../helpers/page"
+import { nextEventNamed, pathname, getFromLocalStorage, setLocalStorageFromEvent } from "../helpers/page"
 
 const path = "/src/tests/fixtures/drive_custom_body.html"
 
@@ -9,15 +9,26 @@ test.beforeEach(async ({ page }) => {
 })
 
 test("test drive with a custom body element", async ({ page }) => {
-  page.click("#drive")
-  await nextBody(page)
+  await page.click("#drive")
+  await nextEventNamed(page, "turbo:load")
 
-  const h1 = await page.locator("h1")
-  const differentContent = await page.locator("#different-content")
+  assert.equal(pathname(page.url()), "/src/tests/fixtures/drive_custom_body_2.html")
+  assert.equal(await page.textContent("h1"), "Drive (with custom body)")
+  assert.equal(await page.textContent("#different-content"), "Drive 2")
+
+  await page.goBack()
+  await nextEventNamed(page, "turbo:load")
+
+  assert.equal(pathname(page.url()), "/src/tests/fixtures/drive_custom_body.html")
+  assert.equal(await page.textContent("h1"), "Drive (with custom body)")
+  assert.equal(await page.textContent("#different-content"), "Drive 1")
+
+  await page.goForward()
+  await nextEventNamed(page, "turbo:load")
 
   assert.equal(pathname(page.url()), "/src/tests/fixtures/drive_custom_body_2.html")
-  assert.equal(await h1.textContent(), "Drive (with custom body)")
-  assert.equal(await differentContent.textContent(), "Drive 2")
+  assert.equal(await page.textContent("h1"), "Drive (with custom body)")
+  assert.equal(await page.textContent("#different-content"), "Drive 2")
 })
 
 test("test drive with mismatched custom body elements", async ({ page }) => {
```

Unfortunately, the test passes without the `await nextAnimationFrame()`. Is it possible that there are other factors involved in your application that might affect the timing in an application-specific way?

Is there something missing from the test that would go along with the `page.goBack()` and `page.goForward()` calls to more faithfully reproduce your use case?

Reply:
I was able to add test coverage that fails with the removal of the `nextAnimationFrame()` call.

Reply:
@seanpdoyle great! What changed in the last ~5 minutes? 😆